OAK-10972 - Do not rely on NumberFormatException when parsing Json Numbers into Doubles

nfsantos opened the pull request:
The current logic first tries to parse numbers as Long, catching a NumberFormatException if that fails and then parsing them as double.
```
try {
    type = PropertyType.LONG;
    values.add(Long.parseLong(number));
} catch (NumberFormatException e) {
    type = PropertyType.DOUBLE;
    values.add(Double.parseDouble(number));
}
```

This is very expensive because the constructor of `Throwable` captures a stack trace. 

This PR changes the parsing logic to not throw an exception when the number is not a long. In a benchmark that iterates over 10 million json strings representing the contents of a typical Oak repository, the implementation in this PR is almost 30% faster:

- current logic: 70s 
- This PR: 50s 

A JFR recording of both runs shows that with the current logic, the JVM raises 127K exceptions per second during the test, while with the logic in this PR the JVM raises less than 1000 exceptions during the whole test execution.

nfsantos committed:
Avoid throwing a NumberFormatException when deserializing Doubles

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into OAK-10972

nfsantos committed:
Replace use of Long.parseLong by ParseUtils.tryParse in parser in DocumentPropertyState.

reschke commented on oak-store-spi/src/main/java/org/apache/jackrabbit/oak/json/ParseUtils.java:
```diff
@@ -0,0 +1,96 @@
+
+        private static final byte[] asciiDigits;
+
+        static {
```
I believe this could be simplified even more, but maybe it doesn't matter a lot.

nfsantos replied:
I do not want to make too many changes to the code, otherwise we risk introducing bugs. Even a change that may seem safe, could introduce non-obvious problems. So unless there is a really good reason to change things, I'd rather leave it close to the original source code.

nit0906 commented on oak-store-spi/src/main/java/org/apache/jackrabbit/oak/json/ParseUtils.java:
```diff
@@ -0,0 +1,96 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.json;
```
this package does not seems fit for this util imo.

in theory, it doesn't really have anything to do with json parsing but rather with string to Long conversion. 

Maybe someplace in oak-core might be more suitable ?

nfsantos replied:
I moved the method to `oak-commons`, in an already existing class `LongUtils`. Initially I was trying to keep the method internal to minimize impact on the public API of Oak, but I think it is better to make it available to other modules in Oak, as this may be useful in other cases.

nit0906 commented on oak-store-spi/src/main/java/org/apache/jackrabbit/oak/json/ParseUtils.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    /*
+     * Simplified version of https://github.com/google/guava/blob/v33.2.1/guava/src/com/google/common/primitives/Longs.java
```
Could we restructure the comment a little - mention what the method does first and then the other details ?

nfsantos replied:
I improved the comments, but as the method is a copy of an external method, I don't think we should copy the full documentation to avoid duplication. I added a link to the original implementation/documentation in Guava and explained the performance reason to use this method instead of the one in the JDK.

nit0906 reviewed (changes requested):
Could we also add a unit test for the new ParseUtils class.

nfsantos committed:
Merge remote-tracking branch 'upstream/trunk' into OAK-10972

nfsantos committed:
Move new method to existing LongUtils class in oak-commons.

nfsantos committed:
Add unit test for new method.

nfsantos commented:
> Could we also add a unit test for the new ParseUtils class.

I added some basic tests, but since the code is a copy from Guava, I don't think it is important to have extensive testing. We should not do any significant modification to the code, so not to introduce bugs.

fabriziofortino commented on oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/LongUtilsTest.java:
missing license

nfsantos committed:
Add missing license

thomasmueller commented on oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/LongUtils.java:
```diff
@@ -51,4 +52,83 @@ public static long safeAdd(long a, long b) {
+     * returns {@code null} instead of throwing an exception if parsing fails. This can be significantly more efficient
+     * when the string to be parsed is often invalid, as raising an exception is an expensive operation.
+     *<p>
+     * This is a simplified version of
```
If we simplify, then I would also hardcode the radix to 10, and get rid of the AsciiDigits class...

nfsantos replied:
I did not look carefully at the algorithm, I just assumed that it is correct, given that it is part of Guava. So I did not want to make more changes than those that seemed totally obvious and safe. And even so, I introduced a bug when replacing the logic for checking for null/empty strings. I agree it would be nice to get rid of AsciiDigits, but I'm not sure if it is worth the effort and risk.

thomasmueller replied:
I think replacing "radix" with a constant RADIX should be simple (given that the value is fixed). That would be a compromise.

nfsantos replied:
Done.

nfsantos replied:
Done, I created a constant, that's cleaner.

thomasmueller commented on oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/LongUtils.java:
```diff
@@ -51,4 +52,83 @@ public static long safeAdd(long a, long b) {
+            return null;
+        }
+        int radix = 10;
+        boolean negative = string.charAt(0) == '-';
```
This does not account for an empty string.

nfsantos replied:
Good catch. Fixed.

thomasmueller commented on oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/LongUtilsTest.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Test
+    public void tryParse() {
+        assertEquals(-1, LongUtils.tryParse("-1").longValue());
```
We need tests for just "-", "-0", and specially for empty string.

thomasmueller replied:
I would add tests cases for ranges "MIN_VALUE .. MIN_VALUE + 100" and "MAX_VALUE - 100 .. MAX_VALUE".

But specially, I would add a lot more tests cases for out-of-range numbers.

nfsantos replied:
Done. Please let me know if there are more tests that should be added. I also changed the tests to compare with `java.util.Long#parseLong`. That is the logic that we are replacing, so the results should be the same.

thomasmueller commented on oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/LongUtils.java:
```diff
@@ -51,4 +52,83 @@ public static long safeAdd(long a, long b) {
+        if (digit < 0 || digit >= radix) {
+            return null;
+        }
+        long accum = -digit;
```
The code is quite smart: it accumulates a negative integer. This is good because there is one more negative number than positive integers. So there won't be an overflow / underflow.

But, there is a risk: due to using negative values, it is harder to understand.

nfsantos replied:
I took the code as a black box, I think we do not need to fully understand it, as long as we do not modify it. This is just a workaround to the restrictions of using Guava directly.

thomasmueller replied:
Well I think we should be able to understand it... if not, then more tests are needed.

thomasmueller commented on oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/LongUtils.java:
```diff
@@ -51,4 +52,83 @@ public static long safeAdd(long a, long b) {
+                return null;
+            }
+            accum *= radix;
+            if (accum < Long.MIN_VALUE + digit) {
```
This line is hard to understand. I think it is correct, but I wouldn't commit it unless if we have sufficient test cases.

nfsantos committed:
Handle the case of the empty string.
Add more test cases.
Test comparing to the result of java.util.Long#parseLong().

thomasmueller commented on oak-commons/src/test/java/org/apache/jackrabbit/oak/commons/LongUtilsTest.java:
```diff
@@ -0,0 +1,72 @@
+
+    private void assertParsingFailure(String value) {
+        try {
+            Long.parseLong(value);
```
Could we verify that the parsed value matches the one from Long.parseLong ?

nfsantos replied:
I'm not sure I understand your suggestion. This should never parse successfully, the test will fail if it does. I improved the error message:
```
    private void assertParsingFailure(String value) {
        try {
            long parsed = Long.parseLong(value);
            fail("Expected NumberFormatException but instead the string: " + value + " was parsed as: " + parsed);
        } catch (NumberFormatException e) {
            // OK
        }
        assertNull(LongUtils.tryParse(value));
    }
```
Is there anything else you think should be done here?

nfsantos committed:
Move radix to a constant.

nfsantos committed:
Improve error message in test.
